### PR TITLE
Spire oidc support

### PR DESCRIPTION
**Pull request by marcofranssen**
This PR is trying to integrate OIDC discovery provider into the Spire chart.

- https://spiffe.io/docs/latest/keyless/vault/readme/#part-1-configure-spire-components
- https://github.com/spiffe/spire-tutorials/tree/master/k8s/oidc-vault

TODO:

- [x] Resolve ACME on localhost deployments (we deploy now soley on http and leave tls to other components like ingress controllers)
- [x] Make it possible to run on http or https (ACME)
- [x] Verify to get a release using this fix https://github.com/spiffe/spire/pull/2398
- [x] Verify to get a release using this fix https://github.com/spiffe/spire/pull/2404 this supersedes 2398

### CODE REVIEW

File: charts/spire/templates/oidc-dp-configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+kind: ConfigMap
+metadata:
+  name: {{ include "spire.fullname" . }}-oidc-discovery-provider
+  namespace: {{ .Release.Namespace }}
+data:
+  oidc-discovery-provider.conf: |
+    log_level = "{{ .Values.oidc.logLevel }}"
+
+    domain = "{{ .Values.oidc.domain }}"
+    acme {
```
Comment (marcofranssen) [inline charts/spire/templates/oidc-dp-configmap.yaml:11]: - We need to figure out how to run TLS on a localhost deployment.
- Would be great if acme could be handled by other components in our infrastructure.
- How to configure the acme challenge to be used e.g. dns challenge

File: charts/spire/templates/server-statefulset.yaml [hunk 2]
```diff
@@ -88,6 +88,31 @@ spec:
...
+          image: "{{ .Values.oidc.image.repository }}:{{ .Values.oidc.image.tag | default .Chart.AppVersion }}"
+          args:
+            - -config
+            - /run/spire/oidc/config/oidc-discovery-provider.conf
+          ports:
+            - containerPort: 443
+              name: spire-oidc-port
+          volumeMounts:
+            - name: spire-server-socket
+              mountPath: {{ include "spire.sockets" . }}
```
Comment (marcofranssen) [inline charts/spire/templates/server-statefulset.yaml:101]: Depending on the booting order of this Pod (3 containers) this socket might not exist yet. 

In the spire-oidc container there is a warning about this.

File: charts/spire/templates/server-configmap.yaml [hunk 3]
```diff
@@ -8,13 +8,16 @@ data:
...
-      socket_path = "{{ include "spire.sockets" . }}/registration.sock"
+      socket_path = "/run/spire/server-sockets/registration.sock"
       trust_domain = {{ .Values.spire.trustDomain | quote }}
       data_dir = "/run/spire/data"
       log_level = "{{ .Values.spire.server.logLevel }}"
       #AWS requires the use of RSA.  EC cryptography is not supported
       ca_key_type = "rsa-2048"
 
+      {{- if eq (.Values.oidc.enabled | toString) "true" }}
+      jwt_issuer = "{{ .Values.oidc.domain }}"
```
Comment (marcofranssen) [inline charts/spire/templates/server-configmap.yaml:19]: The issuer should be retrieved from the `domains` setting now. Maybe we should make it by convention to take the first.

File: charts/spire/templates/oidc-deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+              readOnly: false
+            - name: spire-oidc-config
+              mountPath: /run/spire/oidc/config/
+              readOnly: true
+          readinessProbe:
+            exec:
+              command: ["/bin/ps", "aux", "|", "grep", "oidc-discovery-provider -config /run/spire/oidc/config/oidc-discovery-provider.conf"]
+            initialDelaySeconds: 5
+            periodSeconds: 5
+          readinessProbe:
```
Comment (gjkamstra) [inline charts/spire/templates/oidc-deployment.yaml:60]: Double readinessProbe with same content
Should the second be a health probe?

File: charts/spire/templates/oidc-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+              command: ["/bin/ps", "aux", "|", "grep", "oidc-discovery-provider -config /run/spire/oidc/config/oidc-discovery-provider.conf"]
+            initialDelaySeconds: 5
+            periodSeconds: 5
+          resources:
+            {{- toYaml .Values.resources | nindent 12 }}
+        {{- if .Values.oidc.allowInsecureScheme }}
+        - name: nginx
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: nginx:latest
```
Comment (gjkamstra) [inline charts/spire/templates/oidc-deployment.yaml:71]: Consider making the nginx version configurale in values

File: charts/spire/values.yaml [hunk 6]
```diff
@@ -24,6 +24,42 @@ server:
...
+    type: ClusterIP
+    port: 8081
+
+oidc:
+  enabled: false
+
+  image:
+    repository: gcr.io/spiffe-io/oidc-discovery-provider
+    pullPolicy: IfNotPresent
+    tag: ""
```
Comment (gjkamstra) [inline charts/spire/values.yaml:37]: Should we have a default tag here?

Comment (gjkamstra) [reply charts/spire/values.yaml:37]: Same holds for server above

File: charts/spire/README.md [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 <!-- This README.md is generated. -->
 
-![Version: 0.3.2](https://img.shields.io/badge/Version-0.3.2-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 1.0.0](https://img.shields.io/badge/AppVersion-1.0.0-informational?style=flat-square)
+![Version: 0.4.0](https://img.shields.io/badge/Version-0.4.0-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 1.0.0](https://img.shields.io/badge/AppVersion-1.0.0-informational?style=flat-square)
```
Comment (gjkamstra) [inline charts/spire/README.md:5]: Do we want to document how this chart is able to deploy secure using ACME? 
And insecure using NGINX?

**Review by gjkamstra — COMMENTED** [review]
Looks good.
I added some comments to the review.

File: charts/spire/values.yaml (hunk 6 above)
Comment (marcofranssen) [reply charts/spire/values.yaml:37]: default it takes the version mentioned in Chart.yaml. this allows to override the chart wide appVersion.

File: charts/spire/README.md (hunk 7 above)
Comment (marcofranssen) [reply charts/spire/README.md:5]: These docs are generated, we should indeed document the values.yaml so this becomes more clear from the table in this readme.md

File: charts/spire/templates/oidc-deployment.yaml (hunk 4 above)
Comment (marcofranssen) [reply charts/spire/templates/oidc-deployment.yaml:60]: Good catch. 😊

**Review by gjkamstra — APPROVED** [review]
Great work

**Pull request merged**